lib/Dialect/Arc/ArcFolds.cpp

```diff
@@ -27,3 +27,51 @@ LogicalResult StateOp::canonicalize(StateOp op, PatternRewriter &rewriter) {
 
   return failure();
 }
+
+//===----------------------------------------------------------------------===//
+// MemoryReadPortOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult MemoryReadPortOp::fold(FoldAdaptor adaptor) {
+  if (auto enable = adaptor.getEnable().dyn_cast_or_null<IntegerAttr>()) {
+    if (enable.getValue().isAllOnes()) {
+      this->getEnableMutable().clear();
+      return this->getResult();
+    }
+    if (enable.getValue().isZero()) {
+      // The result is undefined in this case, but we just return 0.
+      return IntegerAttr::get(getType(), 0);
+    }
+  }
+
+  return {};
+}
+
+//===----------------------------------------------------------------------===//
+// Memory write operations
+//===----------------------------------------------------------------------===//
+
+static LogicalResult foldOptionalEnable(Operation *op, Attribute enableAttr,
+                                        MutableOperandRange enableMutable) {
+  if (auto enable = enableAttr.dyn_cast_or_null<IntegerAttr>()) {
+    if (enable.getValue().isAllOnes())
+      enableMutable.clear();
+    if (enable.getValue().isZero())
+      op->erase();
```

Comment:
I believe it's illegal to erase operations in a folder so could you port this into canonicalizer?

Reply:
You're absolutely right, I factored that part of the folder out into a canonicalizer. Thanks for reviewing!